tooling/nargo/src/ops/execute.rs

```diff
@@ -18,6 +18,14 @@ struct ProgramExecutor<'a, B: BlackBoxFunctionSolver, F: ForeignCallExecutor> {
     blackbox_solver: &'a B,
 
     foreign_call_executor: &'a mut F,
+
+    // The Noir compiler codegens per function and call stacks are not shared across ACIR function calls.
+    // We must rebuild a call stack when executing a program of many circuits.
+    call_stack: Vec<ResolvedOpcodeLocation>,
+
+    // Tracks the index of the current function we are executing.
+    // This is used for resolving call stack locations across many function calls.
+    current_function_index: usize,
```

Comment:
With this index we could probably remove passing a borrowed `Circuit` to `execute_circuit` and just fetch it at the beginning of execution, but this felt slightly separate from what this PR wants to do. I could change that here or do it in a follow-up

Reply:
It was a really small change actually so I just did it [84e2c96](https://github.com/noir-lang/noir/pull/4711/commits/84e2c96af7b961024dd5af62499e0a08ade7538d) as part of this PR.